channeldb/invoices.go

```diff
@@ -1999,7 +2029,30 @@ func updateHtlc(resolveTime time.Time, htlc *InvoiceHTLC,
 				"invoice in state canceled")
 		}
 
-	case ContractOpen, ContractAccepted:
+	case ContractAccepted:
+		if htlc.State == HtlcStateSettled {
```

Comment:
@joostjager this is the new section added to handle hodl keysend. the principle here is that for all AMP sets, the preimages of the HTLCs should be known when moving to accept _or_ settle the invoice. the section above validates the transition from open -> settled, while this one handles open -> accepted.

This commit already removed the ability to pass a preimage via the `InvoiceStateUpdateDesc` for AMP HTLC sets. And by validating here, only the `SetID` needs to be passed in when accepting a hodl keysend that has been promoted to AMP. This is consistent for how regular AMP acceptance would look, the preimages and hashes of the HTLCs can be verified before we accept, and the user just needs to pass in the `SetID` of an accepted HTLC set.

Note: could probably use some slight refactoring to reduce code duplication here

Reply:
extracted common code into closure